File: Zeze/Transaction/Collections/PList1.cs
using System;
using System.Collections.Immutable;

namespace Zeze.Transaction.Collections
{
    public sealed class PList1<E> : PList<E>
    {
        public PList1(Func<ImmutableList<E>, Log> logFactory) : base(logFactory)
        {
        }

        public override E this[int index]
        {
            get => Data[index];
            set
            {
                if (value == null)
                {
                    throw new ArgumentNullException("cant add null element");
                }

                if (this.IsManaged)
                {
                    var txn = Transaction.Current;
                    var oldv = txn.GetLog(this) is LogV log ? log.Value : list;
                    txn.PutLog(this, NewLog(oldv.SetItem(index, value)));
                }
                else
                {
                    list = list.SetItem(index, value);
                }
            }
        }

        public override void Add(E item)
        {
            if (item == null)
            {
                throw new ArgumentException("cant add null element");
            }

            if (this.IsManaged)
            {
                var txn = Transaction.Current;
                var oldv = txn.GetLog(this) is LogV log ? log.Value : list;
                txn.PutLog(this, NewLog(oldv.Add(item)));
            }
            else
            {
                list = list.Add(item);
            }
        }

        public override void Clear()
        {
            if (this.IsManaged)
            {
                var txn = Transaction.Current;
                var oldv = txn.GetLog(this) is LogV log ? log.Value : list;
                if (!oldv.IsEmpty)
                {
                    txn.PutLog(this, NewLog(ImmutableList<E>.Empty));
                }
            }
            else
            {
                list = ImmutableList<E>.Empty;
            }
        }


        public override void Insert(int index, E item)
        {
            if (this.IsManaged)
            {
                if (item == null)
                {
                    throw new ArgumentException("cant add null element");
                }

                var txn = Transaction.Current;
                var oldv = txn.GetLog(this) is LogV log ? log.Value : list;
                txn.PutLog(this, NewLog(oldv.Insert(index, item)));
            }
            else
            {
                list = list.Insert(index, item);
            }
        }

        public override bool Remove(E item)
        {
            if (this.IsManaged)
            {
                var txn = Transaction.Current;
                var oldv = txn.GetLog(this) is LogV log ? log.Value : list;
                var newv = oldv.Remove(item);
                if (oldv != newv)
                {
                    txn.PutLog(this, NewLog(newv));
                    return true;
                }
                else
                {
                    return false;
                }
            }
            else
            {
                var newv = list.Remove(item);
                if (newv != list)
                {
                    list = newv;
                    return true;
                }
                else
                {
                    return false;
                }
            }
        }

        public override void RemoveAt(int index)
        {
            if (this.IsManaged)
            {
                var txn = Transaction.Current;
                var oldv = txn.GetLog(this) is LogV log ? log.Value : list;
                txn.PutLog(this, NewLog(oldv.RemoveAt(index)));
            }
            else
            {
                list = list.RemoveAt(index);
            }
        }

        public override void RemoveRange(int index, int count)
        {
            if (this.IsManaged)
            {
                var txn = Transaction.Current;
                var oldv = txn.GetLog(this) is LogV log ? log.Value : list;
                txn.PutLog(this, NewLog(oldv.RemoveRange(index, count)));
            }
            else
            {
                list = list.RemoveRange(index, count);
            }
        }

        protected override void InitChildrenTableKey(TableKey root)
        {

        }
    }
}

File: ZezeJava/src/Zeze/Transaction/Collections/PList1.java
package Zeze.Transaction.Collections;

import org.pcollections.Empty;
import org.pcollections.PVector;

import Zeze.Transaction.*;

import java.util.Collection;

public final class PList1<E> extends PList<E> {
	public PList1(long logKey, LogFactory<PVector<E>> logFactory) {
		super(logKey, logFactory);
	}

	@Override
	public E set(int index, E value) {
		if (value == null) {
			throw new NullPointerException();
		}

		if (this.isManaged()) {
			var txn = Transaction.getCurrent();
			txn.VerifyRecordAccessed(this);
			var log = txn.GetLog(LogKey);
			@SuppressWarnings("unchecked")
			var oldv = null != log ? ((LogV<E>)log).Value : list;
			var olde = oldv.get(index);
			txn.PutLog(NewLog(oldv.with(index, value)));
			return olde;
		}
		else {
			var olde = list.get(index);
			list = list.with(index, value);
			return olde;
		}
	}

	@Override
	public boolean add(E item) {
		if (item == null) {
			throw new NullPointerException();
		}

		if (this.isManaged()) {
			var txn = Transaction.getCurrent();
			txn.VerifyRecordAccessed(this);
			var log = txn.GetLog(LogKey);
			@SuppressWarnings("unchecked")
			var oldv = null != log ? ((LogV<E>)log).Value : list;
			txn.PutLog(NewLog(oldv.plus(item)));
			return true;
		}
		else {
			list = list.plus(item);
			return true;
		}
	}

	@Override
	public boolean addAll(java.util.Collection<? extends E> items) {
		// XXX
		for (var v : items) {
			if (null == v) {
				throw new NullPointerException();
			}
		}

		if (this.isManaged()) {
			var txn = Transaction.getCurrent();
			txn.VerifyRecordAccessed(this);
			var log = txn.GetLog(LogKey);
			@SuppressWarnings("unchecked")
			var oldv = null != log ? ((LogV<E>)log).Value : list;
			txn.PutLog(NewLog(oldv.plusAll(items)));
			return true;
		}
		else {
			list = list.plusAll(items);
			return true;
		}
	}

	@Override
	public void clear() {
		if (this.isManaged()) {
			var txn = Transaction.getCurrent();
			txn.VerifyRecordAccessed(this);
			var log = txn.GetLog(LogKey);
			@SuppressWarnings("unchecked")
			var oldv = null != log ? ((LogV<E>)log).Value : list;
			if (!oldv.isEmpty()) {
				txn.PutLog(NewLog(Empty.vector()));
			}
		}
		else {
			list = Empty.vector();
		}
	}


	@Override
	public void add(int index, E item) {
		if (item == null) {
			throw new NullPointerException();
		}

		if (this.isManaged()) {
			var txn = Transaction.getCurrent();
			txn.VerifyRecordAccessed(this);
			var log = txn.GetLog(LogKey);
			@SuppressWarnings("unchecked")
			var oldv = null != log ? ((LogV<E>)log).Value : list;
			txn.PutLog(NewLog(oldv.plus(index, item)));
		}
		else {
			list = list.plus(index, item);
		}
	}

	@Override
	public boolean removeAll(Collection<?> c) {
		if (this.isManaged()) {
			var txn = Transaction.getCurrent();
			txn.VerifyRecordAccessed(this);
			var log = txn.GetLog(LogKey);
			@SuppressWarnings("unchecked")
			var oldv = null != log ? ((LogV<E>)log).Value : list;
			var newv = oldv.minusAll(c);
			if (oldv != newv) {
				txn.PutLog(NewLog(newv));
				return true;
			}
			else {
				return false;
			}
		}
		else {
			var oldlist = list;
			list = list.minusAll(c);
			return oldlist != list;
		}
	}

	@Override
	public boolean remove(Object item) {
		if (this.isManaged()) {
			var txn = Transaction.getCurrent();
			txn.VerifyRecordAccessed(this);
			var log = txn.GetLog(LogKey);
			@SuppressWarnings("unchecked")
			var oldv = null != log ? ((LogV<E>)log).Value : list;
			var newv = oldv.minus(item);
			if (oldv != newv) {
				txn.PutLog(NewLog(newv));
				return true;
			}
			else {
				return false;
			}
		}
		else {
			var newv = list.minus(item);
			if (newv != list) {
				list = newv;
				return true;
			}
			else {
				return false;
			}
		}
	}

	@Override
	public E remove(int index) {
		if (this.isManaged()) {
			var txn = Transaction.getCurrent();
			txn.VerifyRecordAccessed(this);
			var log = txn.GetLog(LogKey);
			@SuppressWarnings("unchecked")
			var oldv = null != log ? ((LogV<E>)log).Value : list;
			var exist = oldv.get(index);
			txn.PutLog(NewLog(oldv.minus(index)));
			return exist;
		}
		else {
			var exist = list.get(index);
			list = list.minus(index);
			return exist;
		}
	}

	@Override
	protected void InitChildrenRootInfo(Zeze.Transaction.Record.RootInfo root) {

	}
}